rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+
+A sample implementation in a new `TableIndexer` method would look something like
+this:
+
+```py
+def create_and_populate_filtered_index(
+  self, model_name: str, index_suffix: str, **_
+):
+    source_index = f"{model_name}-{index_suffix}"
+    destination_index = f"{source_index}-filtered"
+
+    self.es.indices.create(
+        index=destination_index,
+        body=index_settings(model_name),
+    )
+
+    self.es.reindex(
+        body={
+            "source": {
+                "index": source_index,
+                "query": {
+                    "bool": {
+                        "must_not": [
+                            {
+                                "multi_match": {
+                                    "query": f'"{term}"',
+                                    "fields": ["tags.name", "title", "description"],
+                                }
+                            }
+                            for term in SENSITIVE_TERMS
+                        ]
+                    }
+                },
+            },
+            "dest": {"index": destination_index},
+        },
+        slices="auto",
+        wait_for_completion=True,
+    )
+
+    self.refresh(index_name=destination_index, change_settings=True)
+
+    self.ping_callback()
+```
+
+Please note the following important details:
+
+1. `wait_for_completion` defaults to `True`, but it's worth explicitly including
+   as we rely on the fact that the request will block the process until the
+   reindexing is finished so that we know when we can promote the index.
+1. `slices` is set to `"auto"` so that Elasticsearch is free to decide the
+   optimal number of slices to use to parallelise the reindex.
+1. The source index name is not passed explicitly to the API: instead the method
+   must build it. This means we match the API of other calls (making updates to
+   the data refresh DAG much simpler) and can avoid needing to parse the model
+   name from the source index name for the call to `index_settings`.
+1. We explicitly refresh the index after it is finished reindexing to prime it
+   to be searched. ES only automatically refreshes indexes that are being
+   actively used, and the index should be refreshed before it starts being used.
+
+The creation of the new index follow the same pattern of creating the regular
+index: create the index, copy the data into the index, then issue the command to
+point the alias to the new index. This process is encapsulated in two actions:
+the `create_and_populate_filtered_index` action described above, and the
+existing `point_alias` action.
+
+> **Note**
+>
+> While both processes have a "copy" step, they are notably different. The
+> origin index receives its documents from Postgres via ingestion workers, which
+> push the documents into Elasticsearch. The filtered index receives its
+> documents via the reindex API and does not require the ingestion workers.
```

Comment:
Thanks for pointing out this distinction 🙂 We should make sure that the code path you describe above doesn't spin up the indexer worker instances since they won't be needed.

Reply:
What would be required to ensure that? Assuming we don't call the `schedule_distributed_index` function, it should be fine, right? I don't see anywhere else that would cause the workers to get spun up.

Reply:
I think that's it! Awesome